RavenDB-12725 If journal was not synced then we must not remove it on…

arekpalinski opened the pull request:
… recovery

arekpalinski committed:
RavenDB-12725 If journal was not synced then we must not remove it on recovery

arekpalinski commented:
`CanSurviveRestart` tests is failing - under investigation

ayende committed:
RavenDB-12725
- Fixed a misnamed variable, the journals are not synced here (any longer), just flushed
- Do not delete old journal files at start up, instead, let them be cleaned up normally after the sync

ayende committed:
RavenDB-12725 - Ensure that we will clear the old journals on restart after the next sync

ayende commented:
Okay, I have made some changes, we are not _not_ deleting the files on startup and we are scheduling them to be synced.
Fast tests pass on my machine, but I want to see what the slow tests are doing

ayende commented:
On my end, lots of failing tests:

* SlowTests.Voron.Backups.Incremental.IncrementalBackupShouldCreateConsecutiveJournalFiles
* SlowTests.Voron.Backups.Incremental.CanDoMultipleIncrementalBackupsAndRestoreOneByOne
* SlowTests.Voron.Backups.Incremental.IncrementalBackupShouldCopyJustNewPagesSinceLastBackup

ayende committed:
RavenDB-12725
- Validate the transaction id during restore of incremental backup and ensure that we use the proper starting tx id
- If there are no more files, mark the last synced journal as -1 to avoid trying to start reading from it next time we restart the db

ayende commented:
Hopefully this should fix the tests

arekpalinski commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -76,6 +76,10 @@ public WriteAheadJournal(StorageEnvironment env)
+                if(journalFilesCount == 0)
+                {
+                    header->Journal.LastSyncedJournal = -1;
+                }
```
I'm a bit worried about that change. Note we call `_updateLogInfo` action from `UpdateFileHeaderAfterDataFileSync`

https://github.com/ravendb/ravendb/blob/d476f61ccb2ffbc77fe95772c3599183a1204097/src/Voron/Impl/Journal/WriteAheadJournal.cs#L1216-L1226

and this can override potentially `LastSyncedJournal` we set here:

https://github.com/ravendb/ravendb/blob/d476f61ccb2ffbc77fe95772c3599183a1204097/src/Voron/Impl/Journal/WriteAheadJournal.cs#L1221

BTW The case that `JournalFilesCount == 0` might be tricky for us during recovery. See here:
https://github.com/ravendb/ravendb/blob/d476f61ccb2ffbc77fe95772c3599183a1204097/src/Voron/Impl/Journal/WriteAheadJournal.cs#L163-L167
 I'm gonna double check that - I'm worried that we might still have some journals that were not synced and we should to run the recovery

arekpalinski replied:
Please confirm, but I think you did it specifically for compaction, didn't you? For that purpose I think that would be ok because we force syncing, but might be not to accomplish that so probably only if we `synced` successfully:

https://github.com/ravendb/ravendb/blob/69018d823cb89f4d0bbb9c636b82d14110250f8d/src/Voron/Impl/Compaction/StorageCompaction.cs#L87-L88

arekpalinski committed:
Revert "RavenDB-12725"

This reverts commit d476f61ccb2ffbc77fe95772c3599183a1204097.

arekpalinski committed:
RavenDB-12725 Taking a different approach in addressing the issue. Instead of using LastSyncedJournal during recovery, we'll have _proper_ value in JournalFilesCount that will be the number of journals that we really need in order to successfully do the recovery. This reverts the previous attempt to fix the issue (https://github.com/ravendb/ravendb/pull/8538/commits/ad325b991dfb63f28a9499ae552184ec34bbb06c).

Added tests for backup and the case that we fill up the journal completely and do the restart then.

ayende committed:
RavenDB-12725
- Changing the way we are detecting what range of files we are going to need to recover on database startup
- Making sure that when we sync, we delete the journal files based on the info at the _start_ of the sync operation

WIP - need Voron versioning change

arekpalinski committed:
RavenDB-12725 Fixing incremental backups:
- Validate the transaction id during the restore and ensure that we use the proper starting tx id
- Do not set LastSyncedJournal and CurrentJournal during the restore
- Removing useless WriteAheadJournal.Clear method

arekpalinski committed:
RavenDB-12725 If we delete journal after the successful sync during the compaction operation then we need to set IgnoreMissingLastSyncJournal flag

arekpalinski committed:
RavenDB-12725 Add journal details to message of InvalidJournalException

arekpalinski committed:
RavenDB-12725 Bumping Voron version. Added schema upgrade which applies the change in FileHeader - we changed JournalInfo.JournalFilesCount to JournalInfo.Reserved and JournalInfo.Flags. Test added.

arekpalinski committed:
RavenDB-12725 Let's be prepared for the usage of more flags in the future

arekpalinski committed:
RavenDB-12725 Added test which verifies that we throw on missing journal file during recovery if it was _intentionally_ deleted. Fixed journal path in RavenDB_7099 tests.

ayende commented on src/Voron/Impl/FileHeaders/HeaderAccessor.cs:
```diff
@@ -101,6 +96,28 @@ public bool Initialize()
 
+                if (_theHeader->Version != Constants.CurrentVersion)
+                {
+                    _locker.ExitWriteLock();
```
Why are you exiting the write lock here?

arekpalinski replied:
We're already under write lock so we release it so the updater can access the header accessor. We're in the constructor of storage env so nobody else can touch it anyway.

arekpalinski committed:
RavenDB-12725 Using constant value for number of reserved bytes

arekpalinski committed:
RavenDB-12725 Let's use explicit values for enum flags

arekpalinski committed:
RavenDB-12725 Prevent from potential NRE

arekpalinski commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -934,6 +936,10 @@ private bool GatherInformationToStartSync()
 
                     _lastFlushed = _parent._lastFlushed;
+
+                    if (_lastFlushed.IsValid == false)
```
I got NRE on `_lastFlushed.Journal.SetLastReadTxHeader` (`Journal` was `null`, test `RavenDB_10420`) after I added the suggestion to sync data file after recovery. I did it by adding the following at the end of ctor of `StorageEnvironment`:

```
if (isNew == false && _options.ManualSyncing == false)
	SuggestSyncDataFile(); // let's suggest syncing data file after the recovery
```

The issue was that doing a sync _before_ we flushed anything isn't actually valid because `_lastFlushed` wasn't defined in `UpdateJournalStateUnderWriteTransactionLock`. So I decided to not scheduling the sync immediately after loading a database. We'll do it after the first flush.

ayende replied:
In that case, we need to do a suggested flush, anyway, no?
I don't think it would be triggered until the first commit otherwise, no?

arekpalinski replied:
The recovery on startup is actually the same as flush - it writes data from journals to data file but it's unsynced stuff. So maybe we should set `JournalApplicator._lastFlushed` at the end of the recovery? Like we do at the end of flush?

Then we can safely suggest data sync after the recovery.

ayende replied:
Yes, I think we should set that one.

arekpalinski committed:
RavenDB-12725 Making test more reliable

arekpalinski commented:
All changes squashed in https://github.com/ravendb/ravendb/pull/8556
